### ISSUE TITLE
Inlay hints displayed twice for folded functions/objects

### ISSUE THREAD

**Issue by IllusionMH** [microsoft/vscode#129682]
Issue Type: <b>Bug</b>

Related to #129465 and #129381

Steps to reproduce:
1. Enable `"typescript.inlayHints.variableTypes.enabled": true` and `"typescript.inlayHints.functionLikeReturnTypes.enabled": true` or corresponding JS settings.

2.  Create new TS or JS file and paste next code:
```ts
function test() {
    // code here
}

const obj = {
    a: 'test'
};

function testFolded() {
    // code here
}

const objFolded = {
    a: 'test'
};
```

3. Fold `testFolded` and `objFolded `

![image](https://user-images.githubusercontent.com/1312662/127312985-35882e33-bcb9-488a-84f6-359e855c0dd1.png)

TS Version 4.4.0-dev.20210708

VS Code version: Code - Insiders 1.59.0-insider (e437d424c5e6de97632f5513229c7ebc2cf27a32, 2021-07-28T05:13:47.114Z)
OS version: Windows_NT x64 10.0.19041
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8700 CPU @ 3.20GHz (12 x 3192)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>skia_renderer: enabled_on<br>video_decode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|31.72GB (18.24GB free)|
|Process Argv|--crash-reporter-id cf52dbde-2c94-45c3-9bf3-45d8b9966b04|
|Screen Reader|no|
|VM|0%|
</details>Extensions: none<details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383cf:30185419
pythonvspyt602:30291494
vspor879:30202332
vspor708:30202333
vspor363:30204092
vswsl492cf:30256198
pythonvspyt639:30291487
pythontb:30258533
vspre833:30321513
pythonptprofiler:30281269
vsdfh931cf:30280410
vshan820:30294714
pythondataviewer:30285072
vscus158:30321503
pythonvsuse255:30319630
vscod805cf:30301675
pythonvspyt200:30323110
vscextlangct:30310089
vsccppwt:30312692
pythonvssor306:30340298
bridge0708:30335490

```

</details>

**Comment by mjbvz** [comment]
@jrieken / @hediet Can you please take a look? I believe that folding generates two new requests for inlay hints that VS Code then displays incorrectly. Here's the TS server trace from folding `testFolded` with only `"javascript.inlayHints.functionLikeReturnTypes.enabled": true`:

```
[Trace  - 19:50:26.340] <semantic> Sending request: provideInlayHints (78). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "^untitled:Untitled-1",
    "start": 0,
    "length": 93
}
[Trace  - 19:50:26.341] <semantic> Response received: provideInlayHints (78). Request took 1 ms. Success: true 
Result: [
    {
        "text": ": void",
        "position": {
            "line": 1,
            "offset": 16
        },
        "kind": "Type",
        "whitespaceBefore": true
    },
    {
        "text": ": void",
        "position": {
            "line": 9,
            "offset": 22
        },
        "kind": "Type",
        "whitespaceBefore": true
    }
]
[Trace  - 19:50:26.341] <semantic> Sending request: provideInlayHints (79). Response expected: yes. Current queue length: 0
Arguments: {
    "file": "^untitled:Untitled-1",
    "start": 111,
    "length": 39
}
[Trace  - 19:50:26.342] <semantic> Response received: provideInlayHints (79). Request took 2 ms. Success: true 
Result: [
    {
        "text": ": void",
        "position": {
            "line": 9,
            "offset": 22
        },
        "kind": "Type",
        "whitespaceBefore": true
    }
]
```

The position of the hints seem correct to me. I don't know why VS Code requests the hints twice here though (or why it needs to re-request any hints on folding in this case since no new code became visible)

**Comment by hediet** [comment]
These are the visible ranges:

![image](https://user-images.githubusercontent.com/2931520/128487096-ca6603ad-73a4-4b80-be2b-f5a73f53bd76.png)

And vscode sends two requests, one for each range.
For some reason, each reuqest returns one inlay hint, so in total, there are 2 inlay hints:

![image](https://user-images.githubusercontent.com/2931520/128487168-64845d35-e27f-4930-acfb-3559fc988f5c.png)

@Kingwl do you have an idea why this is?

To me, it looks like a bug in the provider.

**Comment by Kingwl** [comment]
Emmmmmm... I think I know the reason. In the tsserver. For each request, we will get the hint in the area that overlay with the visible range (not strictly included). Honestly, I don't know about this behavior. We may could de-duplicate these items?

**Comment by hediet** [comment]
Can't you just filter out all inlay hints whose position is not in the requested range?

**Comment by Kingwl** [comment]
Well, do you prefer to do it in vscode or ts side?

**Comment by hediet** [comment]
Hmm, maybe actually in vscode.

**Comment by stuartleeks** [comment]
Verified fixed with TypeScript Nightly `v4.5.20210824` in 

```
Version: 1.60.0-insider (user setup)
Commit: e4ac631e9dbf0aa93a4371c4f5f3b8dc43422afb
Date: 2021-08-25T09:43:28.935Z
Electron: 13.1.8
Chrome: 91.0.4472.164
Node.js: 14.16.0
V8: 9.1.269.39-electron.0
OS: Windows_NT x64 10.0.19043
```
